File: languages/go/wordcount.go
// wordcount.go

//
// Count the number of words in a file, and print a list of the top N most
// frequently occurring words.
//
// For words with the same count, sort them by alphabetical order in the
// resulting list.
//
// Convert all uppercase letters A-Z to lowercase a-z, and replace characters
// that are not a-z with spaces.
//

package main

import (
    "fmt"
    "io/ioutil"
    "regexp"
    "sort"
    "strings"
)

func main() {
    //
    // Open the file. ioutil.ReadFile is a utility function that reads all the
    // lines of a file in a slice of bytes.
    //
    fname := "austenPandP.txt"
    // fname := "short.txt"
    bytes, err := ioutil.ReadFile(fname)
    if err != nil {
        panic("Couldn't open file!")
    }

    //
    // Convert letters to lowercase, non-letters to spaces.
    //
    content := strings.ToLower(string(bytes))

    // These two lines convert all non-lowercase letters to spaces.
    rep := regexp.MustCompile(`[^a-z]`)
    processedContent := rep.ReplaceAllString(content, " ")

    //
    // Create an array of all the words.
    //
    // strings.Split returns a slice of strings of all the space-separated
    // words in a string. For example, strings.Split("one two three") returns
    // {"one", "two", "three"}.
    //
    words := strings.Split(processedContent, " ")

    //
    // Remove any extra whitespace at the beginning or end of the strings.
    //
    for i := range words {
        words[i] = strings.TrimSpace(words[i])
    }
 
    //
    // Create a map of the counts of all the words. A map is a hash table. The
    // keys are strings, and the corresponding values are the count of the
    // number of times the string appears.
    //
    // The map is created initially empty. If a string w is not in the map,
    // then freq[w] returns 0, i.e. the zero-value for the type int. This
    // quite convenient in this situation. Many other languages deal with
    // values not in a map by raising an error.
    //
    freq := map[string]int{}
    for _, w := range words {
        if len(w) > 0 {
            freq[w] += 1
        }
    }

    //
    // Create a key-value struct for sorting the frequencies.
    //
    type kv struct {
        key string // a word
        val int    // number of times the word occurs
    }

    //
    // Copy the frequencies into a slice of key-value pairs.
    //
    var arr []kv
    for k, v := range freq {
        arr = append(arr, kv{k, v})
    }

    //
    // We want the result printed in order from most frequent to leat
    // frequent, and for words with the same frequency we want those sorted
    // alphabetically.
    //
    // So we sort twice using a stable sort.
    //
    // First sort alphabetically by word (the key).
    //
    // sort.SliceStable is a standard Go sorting function, and it takes as
    // input a slice (of any type) to sort and a comparison function for
    // comparing two elements in arr.
    cmp := func(i, j int) bool {
        return arr[i].key < arr[j].key
    }
    sort.SliceStable(arr, cmp)

    //
    // Then sort from most frequent to least frequent.
    //
    sort.SliceStable(arr, func(i, j int) bool {
        return arr[i].val > arr[j].val  // note > instead of <
    })

    //
    // Print the top N most frequently occurring words.
    //
    N := 100
    for i, pair := range arr[:N] {
        fmt.Printf("%v. %v (%v)\n", i + 1, pair.key, pair.val)
    }
} // main

File: languages/ruby/wordcount.rb
# wordcount.rb

#
# Count the number of words in a file, and print a list of the top N most
# frequently occurring words. You can print the top N=10 words by default, but
# it should work with any value of N.
#
# For words with same count, sort them by alphabetical order in the list.
#
# Convert all uppercase letters A-Z to lowercase a-z, and replace characters
# that are not a-z with spaces.
#

#
# Create a list of all words in a file.
#
# The file is read as one big string that is first converted to lower case
# using downcase.
# 
# [^a-z] is a regular expression that matches all characters that are *not*
# one of the 26 lowercase letters a-z. Each of those matching characters is
# replaced by a space.
#
# For example:
#
#    "  apple, pear, cherry!!".gsub(/[^a-z]/, ' ')
#    => "  apple  pear  cherry  "
#
# split chops a string into individual words, using spaces as separators. For
# example:
#
#    > "  apple  pear  cherry  ".split
#    => ["apple", "pear", "cherry"]
#
# Altogether, this reads a text file into a list of words consisting of just
# the letters a-z.
#
fname = "austenPandP.txt"
words = File.open(fname).read.downcase.gsub(/[^a-z]/, ' ').split

#
# Create a hash table of (word, count) pairs.
#
# Ruby hash tables begin with { and end with }.
# They consist of key/value pairs, e.g.:
#
#    > fruit = {"pear" => 1, "apple" => 2, "cherry" => 3}
#    > fruit["orange"]
#    => nil
#    > fruit["cherry"]
#    => 3
#
# This codes creates an initially empty hash table, and then adds each words
# from words to it.
#
wc = {}
words.each do |w| 
	if wc.has_key?(w)
		wc[w] += 1
	else
		wc[w] = 1
	end
end

# puts wc

# Sort the pairs first alphabetically by word, and then from highest count to
# lowest count.
sorted_pairs = wc.sort_by {|w,c| w} .sort_by {|w,c| -c }

# Display the list of sorted words and their counts.
puts "Top 10 Most Frequent Words in #{fname}"
sorted_pairs[0..9].each_with_index do |p,i|
	puts "#{i+1}. #{p[0]} (#{p[1]})"
end
